src/devices/SoftwareSpi/SoftwareSpi.cs

```diff
@@ -0,0 +1,199 @@
+        private readonly int _miso;
+        private readonly int _mosi;
+        private readonly int _cs;
+        private readonly SpiConnectionSettings _settings;
+
+        public SoftwareSpi(int clk, int miso, int mosi, int cs, SpiConnectionSettings settings = null, GpioController controller = null)
+        {
+            _controller = controller ?? new GpioController();
+
+            _settings = settings ?? new SpiConnectionSettings(-1, -1);
+
+            _clk = clk;
+            _miso = miso;
+            _mosi = mosi;
+            _cs = cs;
+
+            _controller.OpenPin(_clk, PinMode.Output);
+            _controller.OpenPin(_miso, PinMode.Input);
+            _controller.OpenPin(_mosi, PinMode.Output);
+            _controller.OpenPin(_cs, PinMode.Output);
+
+            // aka. CPOL
+            bool idle = ((int)_settings.Mode & 1) == 1;
+
+            // aka. CPHA
+            bool cpha = ((int)_settings.Mode & 0b10) == 0b10;
+
+            _controller.Write(_cs,!(bool)_settings.ChipSelectLineActiveState);
+            _controller.Write(_clk, idle);
+
+            // TODO: To respect ClockFrequency we need to inject the right delays here
+            //       and have some very accurate way to measure time.
+            //       Ideally we should verify the output with an oscilloscope.
+            if (cpha)
+            {
+                _bitTransfer = new ScopeData(
+                    enter: () => {
+                        _controller.Write(_clk, !idle);
+                    },
+                    exit: () => {
+                        controller.Write(_clk, idle);
+                    });
+            }
+            else
+            {
+                _bitTransfer = new ScopeData(
+                    exit: () => {
+                        _controller.Write(_clk, !idle);
+                        _controller.Write(_clk, idle);
+                    });
+            }
+
+            _chipSelect = new ScopeData(
+                enter: () => {
+                    _controller.Write(_cs, _settings.ChipSelectLineActiveState);
+                },
+                exit: () => {
+                    _controller.Write(_cs, !(bool)_settings.ChipSelectLineActiveState);
+                });
+        }
+
+        public override SpiConnectionSettings ConnectionSettings => _settings;
+
+        public override void TransferFullDuplex(ReadOnlySpan<byte> dataToWrite, Span<byte> dataToRead)
+        {
+            if (dataToRead.Length != dataToRead.Length)
+            {
+                throw new ArgumentException(nameof(dataToWrite));
+            }
+
+            int bitLen = _settings.DataBitLength;
+            int lastBit = bitLen - 1;
+
+            using (ChipSelect)
+            {
+                for (int i = 0; i < dataToRead.Length; i++)
+                {
+                    byte readByte = 0;
+                    for (int j = 0; j < bitLen; j++)
+                    {
+                        using (BitTransfer)
```

Comment:
>BitTransfer [](start = 31, length = 11)

Same question here, where will enter be called?

Reply:
ditto